drivers/ethernet/eth_smsc911x.c

```diff
@@ -0,0 +1,718 @@
+}
+
+int smsc_init(void)
+{
+	unsigned int phyreset = 0;
+
+	if (smsc_check_id() < 0) {
+		return -1;
+	}
+
+	if (smsc_soft_reset() < 0) {
+		return -1;
+	}
+
+	smsc_set_txfifo(5);
+
+	/* Sets automatic flow control thresholds, and backpressure */
+	/* threshold to defaults specified. */
+	SMSC9220->AFC_CFG = 0x006E3740;
+
+#if 0
+	if (smsc_wait_eeprom()) {
+		return -1;
+	}
+#endif
+
+	/* Configure GPIOs as LED outputs. */
+	SMSC9220->GPIO_CFG = 0x70070000;
+
+	smsc_init_irqs();
+
+	/* Configure MAC addresses here if needed. */
+
+	if (smsc_check_phy() < 0) {
+		return -1;
+	}
+
+	if (smsc_reset_phy() < 0) {
+		return -1;
+	}
+
+	k_sleep(PHY_RESET_TIMEOUT);
+	/* Checking whether phy reset completed successfully.*/
+	if (smsc_phy_regread(SMSC9220_PHY_BCONTROL, &phyreset)) {
+		return 1;
+	}
+
+	if (phyreset & (1 << 15)) {
+		return 1;
+	}
+
+	smsc_advertise_caps();
+	/* bit [12] of BCONTROL seems self-clearing. */
+	/* Although it's not so in the manual. */
+	smsc_establish_link();
+
+	/* Interrupt threshold */
+	SMSC9220->FIFO_INT = 0xFF000000;
+
+	smsc_enable_mac_xmit();
+	smsc_enable_xmit();
+	SMSC9220->RX_CFG = 0;
+	smsc_enable_mac_recv();
+
+	/* Rx status FIFO level irq threshold */
+	SMSC9220->FIFO_INT &= ~(0xFF);  /* Clear 2 bottom nibbles */
+
+	/* This sleep is compulsory otherwise txmit/receive will fail. */
+	k_sleep(2000);
+
+	return 0;
+}
+
+/* Driver functions */
+
+static enum ethernet_hw_caps eth_smsc911x_get_capabilities(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
+}
+
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+static struct net_stats_eth *get_stats(struct device *dev)
+{
+	struct eth_context *context = dev->driver_data;
+
+	return &context->stats;
+}
+#endif
+
+static void eth_initialize(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct eth_context *context = dev->driver_data;
+	/* Must be static, netif has pointer stored */
+	static u8_t mac[6];
+
+	LOG_DBG("eth_initialize");
+
+	smsc_read_mac_address(mac);
+
+	SMSC9220->INT_EN |= BIT(SMSC9220_INTERRUPT_RXSTATUS_FIFO_LEVEL);
+
+	net_if_set_link_addr(iface, mac, sizeof(mac), NET_LINK_ETHERNET);
+
+	context->iface = iface;
+
+	ethernet_init(iface);
+}
+
+static int smsc_write_tx_fifo(const u8_t *buf, u32_t len, bool is_last)
+{
+	u32_t *buf32;
+
+	__ASSERT_NO_MSG(((uintptr_t)buf & 3) == 0);
+
+	if (is_last) {
+		/* Last fragment may be not full */
+		len = (len + 3) & ~3;
+	}
+
+	if ((len & 3) != 0 || len == 0) {
+		LOG_ERR("Chunk size not aligned: %u", len);
+		return -1;
+	}
+
+	buf32 = (u32_t *)buf;
+	len /= 4;
+	do {
+		SMSC9220->TX_DATA_PORT = *buf32++;
+	} while (--len);
+
+	return 0;
+}
+
+static int eth_tx(struct device *dev, struct net_pkt *pkt)
+{
+	int res;
+	u16_t total_len = net_pkt_get_len(pkt);
+	u16_t sent = 0;
+	static u8_t tx_buf[1540] __aligned(4);
+	u32_t txcmd_a, txcmd_b;
+	u32_t tx_stat;
+
+	txcmd_a = (1/*is_first_segment*/ << 13) | (1/*is_last_segment*/ << 12)
```

Comment:
You could have created dedicated macros for these bits.

Reply:
> You could have created dedicated macros for these bits.

I could, and actually I converted quite a bunch of raw numbers like that as they were in the original mbedOS. But not everything, and as this PR already spent a lot of time in works, with a lot of side-tracking, I kinda like it to pronounced done finally ;-)